packages/gatsby/src/commands/develop.ts

```diff
@@ -403,26 +404,59 @@ module.exports = async (program: IProgram): Promise<void> => {
     developProcess.send(msg)
   })
 
-  process.on(`beforeExit`, async () => {
-    await Promise.all([
-      watcher?.close(),
-      ...unlocks.map(unlock => unlock()),
-      new Promise(resolve => {
-        statusServer.close(resolve)
-      }),
-      new Promise(resolve => {
-        proxy.server.close(resolve)
-      }),
-    ])
-  })
-
   process.on(`SIGINT`, async () => {
-    await developProcess.stop(`SIGINT`)
-    process.exit()
+    await shutdownServices(
+      {
+        developProcess,
+        unlocks,
+        statusServer,
+        proxy,
+      },
+      `SIGINT`
+    )
+
+    process.exit(0)
   })
 
   process.on(`SIGTERM`, async () => {
-    await developProcess.stop(`SIGTERM`)
-    process.exit()
+    await shutdownServices(
+      {
+        developProcess,
+        unlocks,
+        statusServer,
+        proxy,
+      },
+      `SIGTERM`
+    )
+
+    process.exit(0)
+  })
+
+  onExit((_code, signal) => {
+    shutdownServices(
+      {
+        developProcess,
+        unlocks,
+        statusServer,
+        proxy,
+      },
+      signal as NodeJS.Signals
+    )
+  })
+}
+function shutdownServices(
+  { statusServer, developProcess, proxy, unlocks },
+  signal: NodeJS.Signals
+): Promise<void> {
+  const services = [
+    developProcess.stop(signal),
+    new Promise(resolve => statusServer.close(resolve)),
+    new Promise(resolve => proxy.close(resolve)),
+  ]
+
+  unlocks.forEach(unlock => {
+    services.push(unlock())
   })
+
+  return Promise.all(unlocks).then(() => {})
```

Comment:
```suggestion
  return Promise.all(services).then(() => {})
```

Reply:
me stupid xD

Reply:
Welp - it took me a while to find this too :P